laurat007/assignment2/Ancestors.java

```diff
@@ -0,0 +1,198 @@
+package assignment_2;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+/**
+ *
+ * @author Laura
+ *  This program builds a binary tree by using three methods:
+ *      1. add node in the left extremity
+ *      2. add node in the right extremity
+ *      3. add node in order to balance the tree with the possibility to skip 
+ *      insert locations
+ * 
+ *  The method solving the first task is a recursive one, similar to searching a
+ * node in a tree: when the given value is found, the method will return true, 
+ * but if the value was not found yet, we test if it was found in the current
+ * subtree (in either left or right subtree of the current root) - if this is 
+ * true we add the root to solution and return true, else return false.
+ * 
+ *  The method solving the second task uses the first method as follows:
+ * there are two values given; for each one it finds its ancestors and it 
+ * searches the first common value.
+ * 
+ */
+
+public class Ancestors {
+    static class BinaryTreeNode<T> {
+        BinaryTreeNode left, right;
+        T value;
+        
+        BinaryTreeNode(T value) {
+            this.value  = value;
+        }
+        
+        // add to left extremity
+        void addLeft(T newValue) {
+            BinaryTreeNode aux = this;
+            while(aux.left != null) {
+                aux = aux.left;
+            }
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            aux.left = newNode;
+        }
+        
+        // add to right extremity
+        void addRight(T newValue) {
+            BinaryTreeNode aux = this;
+            while(aux.right != null) {
+                aux = aux.right;
+            }
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            aux.right = newNode;
+        }
+        
+        // add to balance tree
+        // similar to bfs traversal
+        void addBalanced(T newValue, int skips) { 
+            BinaryTreeNode aux = this;
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            Queue<BinaryTreeNode> queue = new LinkedList();
+            queue.add(aux);
+            
+            while(true) {
```

Comment:
Maybe better make it while(!queue.isEmpty()) instead of while(true). It's never good to see while(true) in code... Unless you are writing some service that is supposed to loop indefinitely

Reply:
Done. Thank you for advice! I found it really useful.